extensions/GitExtension/FileExplorerGitIntegration/Models/ThrottledTask.cs

```diff
@@ -0,0 +1,60 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.Diagnostics;
+
+namespace FileExplorerGitIntegration.Models;
+
+internal sealed class ThrottledTask
+{
+    private readonly TimeSpan _interval;
+    private readonly object _lock = new();
+
+    private readonly Action _action;
+
+    private Task? _currentTask;
+    private bool _shouldQueue;
+
+    public ThrottledTask(Action action, TimeSpan interval)
+    {
+        _action = action;
+        _interval = interval;
+    }
+
+    // When the action completes, wait for interval before checking if a new action has been queued.
+    public void Run(CancellationToken cancellationToken = default)
+    {
+        lock (_lock)
+        {
+            if (_currentTask != null && !_currentTask.IsCompleted)

```

Comment:
I'm having a hard time understanding the intent of this method.   

1. If the code has a current task and it is not completed, set _shouldQueue to true and return.  Otherwise,
2. invoke an action.
3. Wait a delay.
4. if _shouldQueue is true, make it false.
5. if _shouldQueue is false, exit the loop and set _currentTask to null.

If _shouldQueue is true, then the action will be invoked twice.  It is invoked on the first pass, _shouldQueue is changed to false, on the second pass the action is invoked again.  Why?  I'm sorry.  Can you explain what this is doing?

Reply:
Going back to read it, you're correct that it's a little gnarly and opaque. Async programins is hard sometimes, everybody! :)

I'm adding comments to clarify the actual intent and expected behavior of this class.